week3/homework/src/main.js

```diff
@@ -0,0 +1,166 @@
+    try {
+      const { error } = Todos.validateTodo(req.body);
+      if (error) return res.status(400).send(error.details.map(d => d.message).join('\n'));
+      const { todo } = req.body;
+      const todos = JSON.parse(await this.readFromFile());
+      const originalTodo = todos.find(todo => todo.id === req.params.id);
+      if (!todo)
+        return res.status(404).send({
+          error: {
+            type: 'Not Found',
+            description: `${req.params.id} is not a valid id`
+          }
+        });
+      originalTodo.description = todo.description;
+      await this.writeToFile(JSON.stringify(todos, null, 2));
+      res.json(originalTodo);
+    }
+ catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async markAsDone(req, res) {
+    try {
+      const todos = JSON.parse(await this.readFromFile());
+      const originalTodo = todos.find(todo => todo.id === req.params.id);
+      if (!originalTodo)
+        return res.status(404).send({
+          error: {
+            type: 'Page Not Found',
+            description: `${req.params.id} is not a valid id`
+          }
+        });
+      originalTodo.done = true;
+      await this.writeToFile(JSON.stringify(todos, null, 2));
+      res.json(originalTodo);
+    }
+ catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async markAsNotDone(req, res) {
+    try {
+      const todos = JSON.parse(await this.readFromFile());
+      const originalTodo = todos.find(todo => todo.id === req.params.id);
+      if (!originalTodo)
+        return res.status(404).send({
+          error: {
+            type: 'Not Found',
+            description: `${req.params.id} is not a valid id`
+          }
+        });
+      originalTodo.done = false;
+      await this.writeToFile(JSON.stringify(todos, null, 2));
+      res.json(originalTodo);
+    }
+ catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async deleteTodo(req, res) {
+    try {
+      const todos = JSON.parse(await this.readFromFile());
+      const deletedTodo = todos.findIndex(todo => todo.id === req.params.id);
+      if (deletedTodo === -1)
+        return res.status(404).send({
+          error: {
+            type: 'Not Found',
+            description: `${req.params.id} is not a valid id`
+          }
+        });
+      todos.splice(deletedTodo, 1);
+      await this.writeToFile(JSON.stringify(todos, null, 2));
+      res.sendStatus(204);
+    }
+ catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async deleteTodos(req, res) {
+    try {
+      await this.writeToFile(JSON.stringify([], null, 2));
+      res.sendStatus(204);
+    }
+ catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  static handleError(res, error) {
+    return res.status(500).send({
+      error: { type: 'Internal Server Error', description: `Error` }
+    });
+  }
+
+  static validateTodo(todo) {
+    const schema = Joi.object().keys({
```

Comment:
interesting approach 👍 
Indeed in real projects we use schemas to validate json instead of manually checking everything using if/else